### DISCUSSION TITLE
Push to restricted master branch

### DISCUSSION THREAD

**Discussion by pgruetter** [community/community#26460] [Actions]
<p>Hi all</p>
<p>I’m trying to set up a release pipeline in GitHub Actions. As part of this, the release number in the Maven pom.xml is set and pushed back to the master branch. </p>
<p>This worked fine when testing in a simple (private) repository. But in the actual repository, we have branch restrictions set on master:</p>
<ul>
<li>
<p>Require pull request reviews before merging (with at least 1 approving review)</p>
</li>
<li>
<p>Restrict who can push to matching branches</p>
</li>
</ul>
<p>The push to master by GitHub Actions fails with</p>
<pre><code>! [remote rejected] HEAD -&gt; master (protected branch hook declined)
</code></pre>
<p>How can I explicitly allow GitHub Actions to push directly to restricted branches? </p>
<p>Found similar discussions (i.e. <a href="https://github.community/t/github-actions-vs-protected-master-branch/17930">here</a>), but the proposed solution of adding github-actions[bot] doesn’t work.</p>
<p>Thanks</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/pgruetter">@pgruetter</a> , </p>
<p>Organization administrators, repository administrators, and users with the Maintain role can always push to a protected branch. </p>
<p>If you are repository administrators , you could use your personal access token to push changes.  You could store it in secrets. Please remember adding persist-credentials: false to checkout@v2.  </p>
<p>There is my workflow yml example for push with PAT: </p>
<pre><code>jobs:
  job1:
    name: push
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
      with:
        persist-credentials: false
    - name: Create local changes
      run: |
       touch $(hexdump -n 16 -v -e '/1 "%02X"' -e '/16 "\n"' /dev/urandom).yml
       git add --all
      working-directory: data/comments/3DACF59A72960405D3B93DF91EE3C060
    - name: Commit files
      run: |
        git config --local user.email "action@github.com"
        git config --local user.name "GitHub Action"
        git commit -m "Add changes " -a

    - name: Push changes
      uses: ad-m/github-push-action@master
      with:
        github_token: ${{ secrets.ADDTAG_PAT }}
        #github_token: ${{ secrets.GITHUB_TOKEN }}
</code></pre>

**Comment by pgruetter** [comment]
<p>Thanks a lot for the precise description and mentioning persist-credentials. It worked!</p>

**Comment by chipzoller** [comment]
<p><span class="mention">@pgruetterI</span>’m an Actions newbie and I’m wanting to do what you did but am not finding the right way. In my case, I simply want to push to the same repo, same branch, but a specific directory. I see the answer which mentioned <a href="https://github.com/ad-m/github-push-action" rel="nofollow noopener">this action</a>, however is this the right thing in my case? Since my workflows are triggered on commit to master, it shouldn’t produce a commit but simply a silent “upload” to that directory and overwrite all contents. Can you maybe give me a little advice?</p>

**Comment by pgruetter** [comment]
<p>Hi <a class="mention" href="/chipzoller">@chipzoller</a> <br>
I don’t fully understand what you’re trying to do. But you can’t just copy files to a directory in git without a commit / push. If you want to add files, you need to checkout your git repo in GitHub Actions, modify it, then commit and push the changes. If that’s what you’re trying to do, then yes, github-push-action is the one I’ve also used here.</p>

**Comment by chipzoller** [comment]
<p>Thanks for your confirmation. I’ll need to look for another way to go about my workflow.</p>
